### ISSUE TITLE
Issues with mail.google.com (Gmail), mobile.twitter.com (Service Worker Chrome bug)

### ISSUE THREAD

**Issue by KushtrimPacaj** [EFForg/privacybadger#2335]
#### What is your browser and browser version?
Google Chrome Version 72.0.3626.121 (Official Build) (64-bit)

#### What is broken and where?
mail.google.com
GMAIL wasn't opening, it was just showing the loading screen, and showing errors in the console "BLOCKED BY CLIENT" stuff.
Tried disabling extensions until I found the one that was breaking gmail, and it was Privacy Badger. Tried enable/disable many times just to be sure.
Here's the output of developer console:
https://imgur.com/a/jSiegav

#### What is the "culprit" domain?
google.com

#### What is your debug output for this domain?
<details><summary>**** ACTION_MAP for google.com</summary>

```
accounts.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
adservice.google.com {
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553628352149,
  "userAction": ""
}
apis.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
books.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
calendar.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
checkout.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
clients1.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
clients2.google.com {
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553500815964,
  "userAction": ""
}
clients6.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
consent.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
contributor.google.com {
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553366216461,
  "userAction": ""
}
cse.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
developers.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
docs.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
drive.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
feedburner.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
feedproxy.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
fusiontables.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
google.com {
  "dnt": false,
  "heuristicAction": "block",
  "nextUpdateTime": 0,
  "userAction": ""
}
groups.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
kh.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
khms0.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
khms1.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
khms2.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
khms3.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
khms4.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
labs.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mail.google.com {
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553563222706,
  "userAction": ""
}
maps-api-ssl.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
maps.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mapsengine.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mt0.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mt1.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mts0.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mts1.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mw1.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
mw2.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
news.google.com {
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553509852973,
  "userAction": ""
}
picasaweb.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
play.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
plus.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
sites.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
smartlock.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
spreadsheets.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
talkgadget.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
translate.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
trends.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
video.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
www.google.com {
  "dnt": false,
  "heuristicAction": "cookieblock",
  "nextUpdateTime": 0,
  "userAction": ""
}
hangouts.google.com {
  "userAction": "",
  "dnt": false,
  "heuristicAction": "block",
  "nextUpdateTime": 1553711495002
}
clients4.google.com {
  "userAction": "",
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553685562434
}
27.client-channel.google.com {
  "userAction": "",
  "dnt": false,
  "heuristicAction": "",
  "nextUpdateTime": 1553879046353
}
```
</details>

**** SNITCH_MAP for google.com
```
google.com [
  "youtube.com",
  "microsoft.com",
  "linkedin.com",
  "solaborate.com"
]
```

**Comment by bcyphers** [comment]
Thanks for filing this. I'm attempting to replicate, but not sure what the problem is yet. What OS are you using?

**Comment by KushtrimPacaj** [comment]
macOS Mojave 10.14.3 (18D109)

**Comment by bcyphers** [comment]
Thanks. Is there any other information you can give us about how you use gmail? Do you usually navigate to the page by typing "mail.google.com" into the browser bar, or by typing "gmail.com" and having it redirect, or some other way? Do you encounter the bug every time you try to use gmail? If not, is there anything you can do to intentionally trigger the bug?

**Comment by KushtrimPacaj** [comment]
Almost always I begin typing "mai...",  and autocomplete to mail.google.com/mail/u/X/#inbox  ( X = {0,1} ; i'm usually logged in two accounts).

I turned off the extension at the moment because of this problem. I'll turn it back on and try to see if there's a pattern, or some way to deterministically trigger the bug. I don't think it happens every time I open gmail.

I'll reply back when I have more info.

**Comment by KushtrimPacaj** [comment]
Did some more testing trying to pinpoint it.
Looks like opening a new tab and writing mail.google.com, the site always works normally. ( based on tonight's testing).
But sometimes if I have a tab open, say wapo.com, and I write mail.google.com on that browser bar and click enter, the problem appears. Doesn't seem to happen for all sites. As far as I could tell, it looked mattered on whether the extension was blocking trackers on that site, but that's just guessing ¯\_(ツ)_/¯.

What I did to kind of consistently reproduce was:

1. Reset tracking domains /Remove all tracking domains on the extension site, so we have a clean slate.
2. Closed && open the browser
3. Opened several sites on different tabs ( in my testings: wapo.com / nytimes.com / telegrafi.com ). Kept opening and closing them.
4. From one of the loaded sites, say wapo.com, navigate to mail.google.com by writing it in the browser bar.

**Comment by drpixie** [comment]
Hi, I'm seeing something similar, noticed it a few days ago (not sure precisely when, google is always frustrating, initially this was just another frustration!). Looks like Privacy Badger is not resetting its idea of the home domain for some URLs.

The address "mail.google.com" works fine. And "https://mail.google.com/mail/u/0" is fine. A bookmark to "https://mail.google.com/mail/u/0/" or "https://mail.google.com/mail/u/0/#inbox" triggers the bug. The URL "https://mail.google.com/mail/u/0/#inbox" is what gets saved when the bookmark is created by "bookmark this page".

Workaround is to change the bookmark to just "mail.google.com".

To reproduce: open a tab, navigate to wikipedia, then enter "https://mail.google.com/mail/u/0/#inbox" or click a bookmark to that address. Gmail fails to fully load (stuck on opening envelope graphic), Privacy Badger shows mail.google.com blocked, and a minute to so later also clients2.google.com as blocked (red slider).

However opening "https://mail.google.com/mail/u/0/#inbox" in a clean (just opened, no content) tab, works just fine and behaves normally.

My system:
Linux Mint 19.1 Cinnamon (Cinnamon 4.0.9) 64-bit
Browser: Chromium 73.0.3683.75 
Privacy Badger: 2019.2.19
Only other extension is flashblock - no change when enable or disabled.

**Comment by ghostwords** [comment]
I wonder if this is related to Service Workers in Chrome (#2327).

**Comment by bcyphers** [comment]
Thank you @drpixie! with that, I'm finally able to reproduce. Another issue I'm encountering (in addition to the "stuck on envelope" problem) is that sometimes Gmail loads, but reports "Oops… the system encountered a problem (#007)" and all asynchronous requests to Google appear to be blocked.

 I'll do some more investigation and report back here.

**Comment by drpixie** [comment]
@bcyphers, I haven't seen the "Oops..." problem, just the "stuck on envelope". I'll advise if anything else comes up.

**Comment by bcyphers** [comment]
I've confirmed that, when you navigate from https://en.wikipedia.org to https://mail.google.com/mail/u/0/#inbox, Privacy Badger still believes the tab is on en.wikipedia.org. Therefore, it interprets all the requests to Google as third-party requests to a known tracker and blocks them, resulting in the freeze.

When I type in `mail.google.com` or https://mail.google.com/mail/u/0/ and hit enter, the first request is a cache hit. Privacy Badger registers this as a main-frame request, updates `tabData` to acknowledge that `mail.google.com` is now the active first party, and proceeds normally. However, when I type in https://mail.google.com/mail/u/0/#inbox, the first request is a Service Worker request, which doesn't register with PB. Therefore I think it's just https://github.com/EFForg/privacybadger/issues/2327 rearing its head again. Unfortunately I don't think there's a way to fix this until the browsers get their act together.

**Comment by KushtrimPacaj** [comment]
Come to think of it, I've seen the "Oops" problem too. It showed a couple of times after I was stuck on loading screen, and clicked the link "Load basic html ( for slow connections )". In the watered down version of gmail, the "oops" message was shown.

To @chrome : https://www.youtube.com/watch?v=xIAfCupuZ3w

**Comment by bcyphers** [comment]
Another update after more poking at it.

Looks like this is different from the [old chromium bug](bugs.chromium.org/p/chromium/issues/detail?id=931588), which is supposed to be fixed. Privacy Badger does intercept some service worker requests, but for some reason doesn't see the initial main frame request. This only happens with the /#inbox URL, but it does happen every time. I've opened a new chromium bug here: https://bugs.chromium.org/p/chromium/issues/detail?id=946908. Still can't reproduce this on Firefox.

**Comment by bcyphers** [comment]
also related to #1997
